lib/Server.js

```diff
@@ -67,6 +66,45 @@ class Server {
 
     this.log = _log || createLogger(options);
 
+    if (this.options.serverMode !== undefined) {
+      this.log.warn(
+        'serverMode is an experimental option, meaning its usage could potentially change without warning'
+      );
+    }
+
+    this.options.serverMode = this.options.serverMode || 'sockjs';
+    let ServerImplementation;
+    let serverImplFound = true;
+    if (typeof this.options.serverMode === 'string') {
+      // could be 'sockjs', in the future 'ws', or a path that should be required
+      if (this.options.serverMode === 'sockjs') {
+        // eslint-disable-next-line global-require
+        ServerImplementation = require('./servers/SockJSServer');
+      } else {
+        try {
+          // eslint-disable-next-line global-require, import/no-dynamic-require
+          ServerImplementation = require(this.options.serverMode);
+        } catch (e) {
+          serverImplFound = false;
+        }
+      }
+    } else if (typeof this.options.serverMode === 'function') {
+      // potentially do more checks here to confirm that the user implemented this properlly
+      // since errors could be difficult to understand
+      ServerImplementation = this.options.serverMode;
+    } else {
+      serverImplFound = false;
+    }
+
+    if (!serverImplFound) {
+      throw new Error(
+        "serverMode must be a string denoting a default implementation (eg. 'sockjs'), a full path to " +
+          'a JS file which exports a class extending BaseServer (webpack-dev-server/lib/servers/BaseServer) ' +
+          'via require.resolve(...), or the class itself which extends BaseServer'
+      );
+    }
+    this.ServerImplementation = ServerImplementation;
```

Comment:
1. Let's move this to util `getSocketServerImplementation`
2. Don't think we need `function`
3. Let's rewrite this on `switch`

Reply:
@evilebottnawi `function` option is good if the user wants to supply an implementation, but their file is exporting multiple things, for example. So like if they want to do:

```
serverMode: require('./my-implementation').Server
```

Reply:
I think better use `require.resolve('./my-implementation/Server')`
/cc @hiroppy

Reply:
I'm fine either way:) Currently, it was fixed as `require.resolve`.